File: Design_Queue.java
// Time Complexity : O(n) where n is the number of integers(operations)
//  worst case is when we push elements continuously and then have to pop all of them onto the output stack.

// Space Complexity : O(n) where n is the number of integers that are being stored at any given time
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :

// Your code here along with comments explaining your approach

class MyQueue {
    // we keep two stacks one for input and output
    private Stack<Integer> input;
    private Stack<Integer> output;

    /** Initialize your data structure here. */
    public MyQueue() {
        input = new Stack<Integer>();
        output = new Stack<Integer>();

    }

    /** Push element x to the back of queue. */
    public void push(int x) {
        // we populate the input stack
        input.push(x);
    }

    /** Removes the element from in front of queue and returns that element. */
    public int pop() {
        // peek reverses the stack so as to form a queue
        peek();
        return output.pop();
    }

    /** Get the front element. */
    public int peek() {
        // here we first see if output is empty..
        if (output.empty())
            // invert the input onto the output so stack behaves like queue
            while (!input.empty()) {
                output.push(input.pop());
            }

        // if output is not empty we dont invert the input to output but just return
        // its peek

        // if output is empty we return the peek of the inverted input stack(which is
        // stored in output stack)
        return output.peek();

    }

    /** Returns whether the queue is empty. */
    public boolean empty() {
        // checks if both the stacks are empty
        return input.empty() && output.empty();
    }
}

/**
 * Your MyQueue object will be instantiated and called as such: MyQueue obj =
 * new MyQueue(); obj.push(x); int param_2 = obj.pop(); int param_3 =
 * obj.peek(); boolean param_4 = obj.empty();
 */

File: Design-queue.py
# Time Complexity : o(N)
# Space Complexity : o(N)
# Did this code successfully run on Leetcode : yes
# Any problem you faced while coding this : no

class MyQueue(object):

    def __init__(self):
        self.stack = []
        self.shuffle_stack = []
        

    def push(self, x):
        """
        Push element x to the back of queue.
        :type x: int
        :rtype: None
        """
        while self.stack:
            self.shuffle_stack.append(self.stack.pop())
            
        self.stack.append(x)
        
        while self.shuffle_stack:
            self.stack.append(self.shuffle_stack.pop())
        

    def pop(self):
        """
        Removes the element from in front of queue and returns that element.
        :rtype: int
        """
        return self.stack.pop()
        

    def peek(self):
        """
        Get the front element.
        :rtype: int
        """
        return self.stack[-1]
        

    def empty(self):
        """
        Returns whether the queue is empty.
        :rtype: bool
        """
        return len(self.stack) == 0

# Your MyQueue object will be instantiated and called as such:
obj = MyQueue()
obj.push(1)
obj.push(2)
param_2 = obj.pop()
param_3 = obj.peek()
param_4 = obj.empty()

print(param_2)
print(param_3)
print(param_4)